### PR TITLE
bpo-42482: remove reference to exc_traceback from TracebackException

### PR DESCRIPTION
**Pull request by iritkatriel**
https://bugs.python.org/issue42482

Automerge-Triggered-By: GH:gvanrossum

### CODE REVIEW

**Review by gvanrossum — APPROVED** [review]
Yeah!

**Comment by iritkatriel** [comment]
> Yeah!

Wait, I made a mistake in the refcount test. The TracebackException object is an intermediate, I need to assign it to something.

**Review by gvanrossum — APPROVED** [review]
Sorry I didn't catch that. :-(

Let me know when you are confident enough that I can merge this.

**Comment by iritkatriel** [comment]
I think it’s ready now.

**Comment by iritkatriel** [comment]
BTW - the faulty test did fail on the old code. It must be that the exc_traceback field was closing a cycle with the test function’s frame or something like that.

**Comment by iritkatriel** [comment]
Actually it might make more sense to move the line that prints the "traceback header" to StackSummary.format(), where the traceback itself is being formatted.

**Comment by gvanrossum** [comment]
> Actually it might make more sense to move the line that prints the "traceback header" to StackSummary.format(), where the traceback itself is being formatted.

Except that would change the behavior of `TracebackException.format()`, which feels like a public(ish) API.

**Comment by iritkatriel** [comment]
> > Actually it might make more sense to move the line that prints the "traceback header" to StackSummary.format(), where the traceback itself is being formatted.
> 
> Except that would change the behavior of `TracebackException.format()`, which feels like a public(ish) API.

Yes, I'm adding a with_header=False age to format().   Alternatively, we could leave the header line where it is but under "if self.stack:" and remove the _print_traceback_header boolean.

**Comment by iritkatriel** [comment]
> BTW - the faulty test did fail on the old code. It must be that the exc_traceback field was closing a cycle with the test function’s frame or something like that.

This is the cycle - the TracebackException was not really intermediate because format() returns a generator:

(Pdb) gc.get_referrers(gc.get_referrers(exc_info[2])[2])
[<traceback.TracebackException object at 0x000001E0D48E3A90>]
(Pdb) gc.get_referrers(gc.get_referrers(gc.get_referrers(exc_info[2])[2])[0])
[<frame at 0x000001E0D2551C90, file 'C:\\Users\\User\\src\\cpython\\lib\\traceback.py', line 616, code format>]
(Pdb) gc.get_referrers(gc.get_referrers(gc.get_referrers(gc.get_referrers(exc_info[2])[2])[0])[0])
[<generator object TracebackException.format at 0x000001E0D49E63C0>]
(Pdb) gc.get_referrers(gc.get_referrers(gc.get_referrers(gc.get_referrers(gc.get_referrers(exc_info[2])[2])[0])[0])[0])
[{'self': <test.test_traceback.TestTracebackException testMethod=test_no_refs_to_exception_and_traceback_objects>, 'exc_info': (<class 'ZeroDivisionError'>, ZeroDivisionError('division by zero'), <traceback object at 0x000001E0D48DE880>), 'refcnt1': 2, 'refcnt2': 3, **'_': <generator object TracebackException.format at 0x000001E0D49E63C0>**, 'gc': <module 'gc' (built-in)>}]

**Comment by iritkatriel** [comment]
I went for just removing the boolean. At this point moving the header line (as an opt-in) won't make the code clearer than this.

File: Lib/traceback.py [hunk 1]
```diff
@@ -627,7 +626,7 @@ def format(self, *, chain=True):
                 not self.__suppress_context__):
                 yield from self.__context__.format(chain=chain)
                 yield _context_message
-        if self._print_traceback_header:
+        if self.stack:
             yield 'Traceback (most recent call last):\n'
         yield from self.stack.format()
```
Comment (gvanrossum) [inline Lib/traceback.py]: Maybe indent this line too then?

**Review by gvanrossum — COMMENTED** [review]
Okay, though I am curious why the original fixer didn't do it that way...

**Review by gvanrossum — APPROVED** [review]
Let's sit on this for a day...

**Pull request merged**